README
***************************************************************************
>>> What it does
***************************************************************************
Deploy an Intel microcode. This tool is obsolete and the microcode is the
subject to be distributed via kernel-firmware, however Intel still does not
supply the microcode in a form consumable by the Linux's microcode driver. So
that this tool transform Intel's microcode as well as deploy it.

***************************************************************************
>>> Contributors
***************************************************************************
Credit where credit is due; the following people contributed ideas or
patches. Thanks :)

Kay Sievers <kay.sievers@vrfy.org>

***************************************************************************
>>> Special thanks
***************************************************************************
o Intel Corporation, for supplying microcode update data and publishing the
  specifications that enabled us to write microcode driver for Linux.

- Anton Arapov <aarapov@redhat.com> https://pagure.io/microcode_ctl/


Changelog
2.1-22 20 Jun 2019, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20190618

2.1-21 15 May 2019, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20190514

2.1-20 09 May 2019, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20190312

2.1-19 13 August 2018, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20180807

2.1-18 09 July 2018, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20180703

2.1-17 16 May 2018, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20180425

2.1-16 15 March 2018, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20180312

2.1-15 09 January 2018, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20180108

2.1-14 21 November 2017, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20171117

2.1-13 12 July 2017, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20170707

2.1-12 23 May 2017, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20170511

2.1-11 02 December 2016, Anton Arapov <aarapov@redhat.com>
      Intel CPU microcode update. 20161104

2.1-10 21 July 2016, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20160714

2.1-9 24 June 2016, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20160607

2.1-8 12 January 2016, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20151106

2.1-7 03 February 2015, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20150121

2.1-6 21 September 2014, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20140913

2.1-5 08 July 2014, Anton Arapov <arapov@gmail.com>
      Intel CPU microcode update. 20140624

2.1-4 01 May 2014, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20140430

2.1-3 23 January 2014, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20140122

2.1-2 09 September 2013, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20130906

2.1-1 14 August 2013, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20130808

2.1   27 July 2013, Anton Arapov <anton@redhat.com>
      Remove AMD CPU microcode. Moved to linux-firmware.

2.0-3 03 April 2013, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20130222

2.0-2 17 October 2012, Anton Arapov <anton@redhat.com>
      AMD CPU microcode update. 2012-09-10

2.0-1 02 October 2012, Anton Arapov <anton@redhat.com>
      Intel CPU microcode update. 20120606-v2

2.0   25 July 2012, Anton Arapov <anton@redhat.com>
      Initial release.

--
homepage: http://fedorahosted.org/microcode_ctl/
old releases archive: https://fedorahosted.org/released/microcode_ctl


microcode_ctl.c
/*
 * microcode_ctl - Manipulate /dev/cpu/microcode under Linux
 *
 * Copyright 2000 (c) Simon Trimmer, Tigran Aivazian.
 * Copyright 2012 (c) Anton Arapov.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version
 * 2 of the License, or (at your option) any later version.
 *
 */

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <string.h>

static char *progname;
int print_normal_messages=1;
int print_error_messages=1;

#define BUFFER_SIZE	4096
#define MAX_MICROCODE	2000000

#define MICROCODE_DEVICE_DEFAULT	"/dev/cpu/microcode"
#define MICROCODE_FILE_DEFAULT		"/lib/firmware/microcode.dat"

static void usage(void)
{
	fprintf(stderr, "\nThis program is for updating the microcode on Intel processors\n"
			"belonging to the IA32 family - PentiumPro upwards (x86_64 included).\n"
			"It depends on the Linux kernel microcode driver.\n\n"
			"Usage: %s [-h] [-u] [-q] [-Q] [-f microcode]\n\n"
			"  -h	this usage message\n"
			"  -q	run silently when successful\n"
			"  -Q	run silently even on failure\n"
			"  -u	upload microcode (default filename:\"%s\"\n"
			"  -f	upload microcode from named Intel formatted file\n\n", 
			progname, MICROCODE_FILE_DEFAULT);
	exit(1);
}

/* 
 * The update has two stages; 
 * a) read in the Intel microcode file and convert it into a format suitable
 *    for the processor.
 * b) send the microcode to the driver which applies the update
 */
static int do_update(char *device, char *filename)
{
	FILE *fp;
	char line_buffer[BUFFER_SIZE];
	int *microcode;
	int *pos;
	int outfd;
	int wrote, length;

	microcode=malloc(sizeof(int)* MAX_MICROCODE);
	if (!microcode) return ENOMEM;

	if( (fp=fopen(filename, "r")) == NULL){
		if(print_error_messages)
			fprintf(stderr, "%s: cannot open source file '%s' errno=%d (%s)\n",
				progname, filename, errno, strerror(errno));
		goto end;
	}

	pos = microcode;

	while(fgets(line_buffer, BUFFER_SIZE, fp) != NULL) {
		 /*
		  * Data lines will are of the form "%x, %x, %x, %x", therefore
		  * lines start with a 0
		  */
		if(*line_buffer == '0'){
			sscanf(line_buffer, "%x, %x, %x, %x", pos,
					(pos + 1), (pos + 2), (pos + 3));
			pos += 4;
		}

		if (MAX_MICROCODE < (pos - microcode)){
			/* not checking the buffer length could cause grief? */
			if(print_error_messages)
				fprintf(stderr, "%s: file too large for utility microcode buffer\n"
						"%s: change MAX_MICROCODE yourself :)\n", progname, progname);
			fclose(fp);	
			goto end;
		}
		
	}

	fclose(fp);
	length = sizeof(int) * (pos - microcode);
	if(print_normal_messages)
		fprintf(stderr, "%s: writing microcode (length: %d)\n",  progname, length);
	
	if((outfd = open(device, O_WRONLY)) == -1){
		if(print_error_messages)
			fprintf(stderr, "%s: cannot open %s for writing errno=%d (%s)\n",
				progname, device, errno, strerror(errno));
		goto end;
	}

	if( (wrote = write(outfd, microcode, length)) < 0){
		if(print_error_messages)
			fprintf(stderr, "%s: error writing to '%s' errno=%d (%s)\n"
					"%s: there may be messages from the driver in your system log.\n",
				progname, device, errno, strerror(errno), progname);
		close(outfd);
		goto end;
	}

	if((wrote == length) && print_normal_messages)
			fprintf(stderr, "%s: microcode successfuly written to %s\n",
			       progname, device);

	close(outfd);

 end:
	free(microcode);
	return errno;
}

int main(int argc, char *argv[])
{
	int c;
	static char device[2048];
	static char filename[2048];
	int upload=0;
	int return_code;

	progname = argv[0];

	strcpy(device, MICROCODE_DEVICE_DEFAULT);
	strcpy(filename, MICROCODE_FILE_DEFAULT);
	
	while (EOF != (c = getopt(argc, argv, "hqQud:f:"))) {
		switch(c) {
			case 'h':
				usage();

			case 'q':
				print_normal_messages=0;
				break;

			case 'Q':
				print_error_messages=0;
				print_normal_messages=0;
				break;

			case 'd':
				strncpy(device, optarg, sizeof(device));
				device[sizeof(device)-1] = '\0'; /* ensure null terminated */
				break;

			case 'u': /* do a microcode upload */
				upload++;
				break;

			case 'f': /* set microcode file to optarg and upload */
				upload++;
				strncpy(filename, optarg, sizeof(filename));
				filename[sizeof(filename)-1] = '\0'; /* ensure null terminated */
				break;

			case '?':
				usage();
		}
	}

	if (upload) {
		if((return_code = do_update(device, filename)))
			exit(return_code);
	} else
		usage();

	return 0;
}


intel-microcode2ucode.c
/*
 * Convert Intel microcode.dat into individual ucode files
 * named: intel-ucode/$family-$model-$stepping
 *
 * The subdir intel-ucode/ is created in the current working
 * directory. We get multiple ucodes in the same file, so they
 * are appended to an existing file. Make sure the directory
 * is empty before every run of the converter.
 *
 * Kay Sievers <kay.sievers@vrfy.org>
 * Anton Arapov <anton@redhat.com>
 */


#ifndef _GNU_SOURCE
# define _GNU_SOURCE 1
#endif

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <limits.h>
#include <stdbool.h>
#include <inttypes.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/stat.h>

struct microcode_header_intel {
	unsigned int hdrver;
	unsigned int rev;
	unsigned int date;
	unsigned int sig;
	unsigned int cksum;
	unsigned int ldrver;
	unsigned int pf;
	unsigned int datasize;
	unsigned int totalsize;
	unsigned int reserved[3];
};

union mcbuf {
	struct microcode_header_intel hdr;
	unsigned int i[0];
	char c[0];
};

int main(int argc, char *argv[])
{
	char *filename = "/lib/firmware/microcode.dat";
	FILE *input, *f;
	char line[LINE_MAX];
	char buf[4000000];
	union mcbuf *mc;
	size_t bufsize, count, start;
	int rc = EXIT_SUCCESS;

	if (argv[1] != NULL)
		filename = argv[1];

	if (!strcmp(filename, "-")) {
		input = stdin;
	} else {
		input = fopen(filename, "re");
		if (input == NULL) {
			printf("open %s: %m\n", filename);
			rc = EXIT_FAILURE;
			goto out;
		}
	}

	count = 0;
	mc = (union mcbuf *) buf;
	while (fgets(line, sizeof(line), input) != NULL) {
		if (sscanf(line, "%x, %x, %x, %x",
		    &mc->i[count],
		    &mc->i[count + 1],
		    &mc->i[count + 2],
		    &mc->i[count + 3]) != 4)
			continue;
		count += 4;
	}
	fclose(input);

	bufsize = count * sizeof(int);
	printf("%s: %lu(%luk) bytes, %zu integers\n",
	       filename,
	       bufsize,
	       bufsize / 1024,
	       count);

	if (bufsize < sizeof(struct microcode_header_intel))
		goto out;

	mkdir("intel-ucode", 0750);

	start = 0;
	for (;;) {
		size_t size;
		unsigned int family, model, stepping;
		unsigned int year, month, day;

		mc = (union mcbuf *) &buf[start];

		if (mc->hdr.totalsize)
			size = mc->hdr.totalsize;
		else
			size = 2000 + sizeof(struct microcode_header_intel);

		if (mc->hdr.ldrver != 1 || mc->hdr.hdrver != 1) {
			printf("unknown version/format:\n");
			rc = EXIT_FAILURE;
			break;
		}

		/*
		 *  0- 3 stepping
		 *  4- 7 model
		 *  8-11 family
		 * 12-13 type
		 * 16-19 extended model
		 * 20-27 extended family
		 */
		family = (mc->hdr.sig >> 8) & 0xf;
		if (family == 0xf)
			family += (mc->hdr.sig >> 20) & 0xff;
		model = (mc->hdr.sig >> 4) & 0x0f;
		if (family == 0x06)
			model += ((mc->hdr.sig >> 16) & 0x0f) << 4;
		stepping = mc->hdr.sig & 0x0f;

		year = mc->hdr.date & 0xffff;
		month = mc->hdr.date >> 24;
		day = (mc->hdr.date >> 16) & 0xff;

		asprintf(&filename, "intel-ucode/%02x-%02x-%02x", family, model, stepping);
		printf("\n");
		printf("%s\n", filename);
		printf("signature: 0x%02x\n", mc->hdr.sig);
		printf("flags:     0x%02x\n", mc->hdr.pf);
		printf("revision:  0x%02x\n", mc->hdr.rev);
		printf("date:      %04x-%02x-%02x\n", year, month, day);
		printf("size:      %zu\n", size);

		f = fopen(filename, "ae");
		if (f == NULL) {
			printf("open %s: %m\n", filename);
			rc = EXIT_FAILURE;
			goto out;
		}
		if (fwrite(mc, size, 1, f) != 1) {
			printf("write %s: %m\n", filename);
			rc = EXIT_FAILURE;
			goto out;
		}
		fclose(f);
		free(filename);

		start += size;
		if (start >= bufsize)
			break;
	}
	printf("\n");

 out:
	return rc;
}
